src/dataframerow/utils.jl

```diff
@@ -167,43 +185,83 @@ function row_group_slots(cols::NTuple{N,<:Union{CategoricalVector,PooledVector}}
     # but it needs to remain reasonable compared with the size of the data frame.
     if prod(Int128.(ngroupstup)) > typemax(Int) || ngroups > 2 * length(groups)
         return invoke(row_group_slots,
-                      Tuple{Tuple{Vararg{AbstractVector}}, Val,
-                            Union{Vector{Int}, Nothing}, Bool},
-                      cols, hash, groups, skipmissing)
+                      Tuple{Tuple{Vararg{AbstractVector}}, Any, Val,
+                            Union{Vector{Int}, Nothing}, Bool, Bool},
+                      cols, refpools, hash, groups, skipmissing, sort)
     end
 
     seen = fill(false, ngroups)
-    # Compute vector mapping missing to -1 if skipmissing=true
-    refmaps = map(cols) do col
-        nlevs = nlevels(col)
-        refmap = collect(-1:(nlevs-1))
-        # First value in refmap is only used by CategoricalArray
-        # (corresponds to ref 0, i.e. missing values)
-        refmap[1] = skipmissing ? -1 : nlevs
-        if col isa PooledArray{>: Missing} && skipmissing
-            missingind = get(col.invpool, missing, 0)
-            if missingind > 0
-                refmap[missingind+1] = -1
-                refmap[missingind+2:end] .-= 1
+    refs = map(DataAPI.refarray, cols)
+    strides = (cumprod(collect(reverse(ngroupstup)))[end-1:-1:1]..., 1)::NTuple{N,Int}
+    firstinds = map(firstindex, refpools)
+    # TODO: when skipmissing=true, do not include missing values
+    # when checking whether pool is sorted
+    if sort && !all(issorted, refpools)
+        # Compute vector mapping missing to -1 if skipmissing=true
+        refmaps = map(cols, refpools) do col, refpool
+            refmap = collect(0:length(refpool)-1)
+            if skipmissing
+                fi = firstindex(refpool)
+                missingind = findfirst(ismissing, refpool)
+                if missingind !== nothing
+                    mi = something(missingind)
+                    refmap[mi-fi+1] = -1
+                    refmap[mi-fi+2:end] .-= 1
+                end
+                if sort
+                    nm = missingind === nothing ? eachindex(refpool) :
+                        setdiff(eachindex(refpool), missingind)
+                    perm = sortperm(view(refpool, nm))
+                    invpermute!(view(refmap, nm .- fi .+ 1), perm)
+                end
+            elseif sort
+                # FIXME: collect is needed for CategoricalRefPool
+                perm = sortperm(collect(refpool))
+                invpermute!(refmap, perm)
             end
+            refmap
         end
-        refmap
-    end
-    strides = (cumprod(collect(reverse(ngroupstup)))[end-1:-1:1]..., 1)::NTuple{N,Int}
-    @inbounds for i in eachindex(groups)
-        local refs
-        let i=i # Workaround for julia#15276
-            refs = map(c -> c.refs[i], cols)
+        @inbounds for i in eachindex(groups)
+            local refs_i
+            let i=i # Workaround for julia#15276
```

Comment:
as discussed `getindex.(refs, i)` should be a bit faster I think.

Reply:
I just tested it, it doesn't make any difference here. That's probably because the loop does many different things.

For reference, I've also tried using a generated function, but the result isn't faster (actually even a bit slower). SIMD isn't used unless I remove the `seen[gid] = true` line (note that without the generated function we don't get SIMD even without it). I kind of expected this. Apparently scatter instructions are available on AVX512, but I don't have access to such a machine.

What's weird is that remove `seen` makes things a bit *slower* on my machine (Skylake with AVX2), despite (or because?) the use of SIMD. Maybe this code is memory-bound so SIMD doesn't make it any faster? (If removing `seen` was faster, we could have kept track of the number of groups we have seen, and when we have seen all of them switch to a loop without it.)

```julia
@generated function loop_helper(groups::AbstractVector{<: Integer},
                                seen::AbstractVector{Bool},
                                refs::NTuple{N, <: AbstractVector{<: Integer}},
                                strides::NTuple{N, Integer},
                                firstinds::NTuple{N, Integer},
                                missinginds::NTuple{N, Integer},
                                skipmissing::Bool) where N
    ex = :()
    for j in 1:N
        ex = :(begin
            $ex
            r = Int(refs[$j][i])
            missingind = missinginds[$j]
            if skipmissing
                if r == missingind
                    r = -1
                    anymissing = true
                elseif r > missingind
                    r -= 1
                end
            end
            fi = firstinds[$j]
            gid += (r-fi) * strides[$j]
        end)
    end
    quote
        @inbounds for i in eachindex(groups)
            gid = 1
            anymissing = false
            $ex
            if skipmissing && anymissing
                gid = 0
            else
                seen[gid] = true
            end
            groups[i] = gid
        end
    end
end
```

Reply:
Then let us have it as is for now. I think *skylake* is currently what we should optimize for. Thank you for looking into it.

Reply:
I could check, and even with a version of Julia built from source to target the native CPU, I don't get SIMD on a machine that supports AVX512.